File: backend/src/main/java/io/github/amandajuchem/projetoapi/services/FacadeService.java
package io.github.amandajuchem.projetoapi.services;

import io.github.amandajuchem.projetoapi.entities.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Service;

import java.util.UUID;

/**
 * The type Facade service.
 */
@Service
@RequiredArgsConstructor
public class FacadeService {

    private final AdocaoService adocaoService;
    private final AgendamentoService agendamentoService;
    private final AnimalService animalService;
    private final AtendimentoService atendimentoService;
    private final EnderecoService enderecoService;
    private final ExameService exameService;
    private final FichaMedicaService fichaMedicaService;
    private final ImagemService imagemService;
    private final TelefoneService telefoneService;
    private final TutorService tutorService;
    private final UsuarioService usuarioService;

    //////////////////////////////////////////////// ADOÇÃO ////////////////////////////////////////////////

    /**
     * Delete adoção.
     *
     * @param id the id
     */
    public void adocaoDelete(UUID id) {
        adocaoService.delete(id);
    }

    /**
     * Find all adoções.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of adoção
     */
    public Page<Adocao> adocaoFindAll(Integer page, Integer size, String sort, String direction) {
        return adocaoService.findAll(page, size, sort, direction);
    }

    /**
     * Find adoção by id.
     *
     * @param id the id
     * @return the adoção
     */
    public Adocao adocaoFindById(UUID id) {
        return adocaoService.findById(id);
    }

    /**
     * Save adoção.
     *
     * @param adocao the adocao
     * @return the adoção
     */
    public Adocao adocaoSave(Adocao adocao) {
        return adocaoService.save(adocao);
    }

    //////////////////////////////////////////////// AGENDAMENTO ////////////////////////////////////////////////

    /**
     * Delete agendamento.
     *
     * @param id the id
     */
    public void agendamentoDelete(UUID id) {
        agendamentoService.delete(id);
    }

    /**
     * Find all agendamentos.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list
     */
    public Page<Agendamento> agendamentoFindAll(Integer page, Integer size, String sort, String direction) {
        return agendamentoService.findAll(page, size, sort, direction);
    }

    /**
     * Save agendamento.
     *
     * @param agendamento the agendamento
     * @return the agendamento
     */
    public Agendamento agendamentoSave(Agendamento agendamento) {
        return agendamentoService.save(agendamento);
    }

    /**
     * Search agendamentos.
     *
     * @param value     Data, nome do animal ou nome do veterinário
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of agendamentos
     */
    public Page<Agendamento> agendamentoSearch(String value, Integer page, Integer size, String sort, String direction) {
        return agendamentoService.search(value, page, size, sort, direction);
    }

    //////////////////////////////////////////////// ANIMAL ////////////////////////////////////////////////

    /**
     * Delete animal.
     *
     * @param id the id
     */
    public void animalDelete(UUID id) {
        animalService.delete(id);
    }

    /**
     * Find all animais.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list
     */
    public Page<Animal> animalFindAll(Integer page, Integer size, String sort, String direction) {
        return animalService.findAll(page, size, sort, direction);
    }

    /**
     * Find animal by id.
     *
     * @param id the id
     * @return the animal
     */
    public Animal animalFindById(UUID id) {
        return animalService.findById(id);
    }

    /**
     * Save animal.
     *
     * @param animal the animal
     * @return the animal
     */
    public Animal animalSave(Animal animal) {
        return animalService.save(animal);
    }

    /**
     * Search animais.
     *
     * @param value     Nome
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of animais
     */
    public Page<Animal> animalSearch(String value, Integer page, Integer size, String sort, String direction) {
        return animalService.search(value, page, size, sort, direction);
    }

    //////////////////////////////////////////////// ATENDIMENTO ////////////////////////////////////////////////

    /**
     * Delete atendimento.
     *
     * @param id the id
     */
    public void atendimentoDelete(UUID id) {
        atendimentoService.delete(id);
    }

    /**
     * Find all atendimentos.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list
     */
    public Page<Atendimento> atendimentoFindAll(Integer page, Integer size, String sort, String direction) {
        return atendimentoService.findAll(page, size, sort, direction);
    }

    /**
     * Find atendimento by id.
     *
     * @param id the id
     * @return the atendimento
     */
    public Atendimento atendimentoFindById(UUID id) {
        return atendimentoService.findById(id);
    }

    /**
     * Save atendimento.
     *
     * @param atendimento the atendimento
     * @return the atendimento
     */
    public Atendimento atendimentoSave(Atendimento atendimento) {
        return atendimentoService.save(atendimento);
    }


    /**
     * Search atendimentos.
     *
     * @param value     Data, nome do animal ou nome do veterinário
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of atendimentos
     */
    public Page<Atendimento> atendimentoSearch(String value, Integer page, Integer size, String sort, String direction) {
        return atendimentoService.search(value, page, size, sort, direction);
    }

    //////////////////////////////////////////////// ENDEREÇO ////////////////////////////////////////////////

    /**
     * Delete endereco.
     *
     * @param id the id
     */
    public void enderecoDelete(UUID id) {
        enderecoService.delete(id);
    }

    /**
     * Save endereco.
     *
     * @param endereco the endereco
     * @return the endereco
     */
    public Endereco enderecoSave(Endereco endereco) {
        return enderecoService.save(endereco);
    }

    //////////////////////////////////////////////// EXAME ////////////////////////////////////////////////

    /**
     * Delete exame.
     *
     * @param id the id
     */
    public void exameDelete(UUID id) {
        exameService.delete(id);
    }

    /**
     * Find all exames.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list
     */
    public Page<Exame> exameFindAll(Integer page, Integer size, String sort, String direction) {
        return exameService.findAll(page, size, sort, direction);
    }

    /**
     * Save exame.
     *
     * @param exame the exame
     * @return the exame
     */
    public Exame exameSave(Exame exame) {
        return exameService.save(exame);
    }

    /**
     * Search exames.
     *
     * @param value     Nome ou categoria
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the page
     */
    public Page<Exame> exameSearch(String value, Integer page, Integer size, String sort, String direction) {
        return exameService.search(value, page, size, sort, direction);
    }

    //////////////////////////////////////////////// FICHA MÉDICA ////////////////////////////////////////////////

    /**
     * Delete ficha medica.
     *
     * @param id the id
     */
    public void fichaMedicaDelete(UUID id) {
        fichaMedicaService.delete(id);
    }

    /**
     * Save ficha medica.
     *
     * @param fichaMedica the ficha medica
     * @return the ficha medica
     */
    public FichaMedica fichaMedicaSave(FichaMedica fichaMedica) {
        return fichaMedicaService.save(fichaMedica);
    }

    //////////////////////////////////////////////// IMAGEM ////////////////////////////////////////////////

    /**
     * Delete imagem.
     *
     * @param id the id
     */
    public void imagemDelete(UUID id) {
        imagemService.delete(id);
    }

    /**
     * Save imagem.
     *
     * @param imagem the imagem
     * @return the imagem
     */
    public Imagem imagemSave(Imagem imagem) {
        return imagemService.save(imagem);
    }

    //////////////////////////////////////////////// TELEFONE ////////////////////////////////////////////////

    /**
     * Delete telefone.
     *
     * @param id the id
     */
    public void telefoneDelete(UUID id) {
        telefoneService.delete(id);
    }

    /**
     * Save telefone.
     *
     * @param telefone the telefone
     * @return the telefone
     */
    public Telefone telefoneSave(Telefone telefone) {
        return telefoneService.save(telefone);
    }

    //////////////////////////////////////////////// TUTOR ////////////////////////////////////////////////

    /**
     * Delete tutor.
     *
     * @param id the id
     */
    public void tutorDelete(UUID id) {
        tutorService.delete(id);
    }

    /**
     * Find all tutores.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of tutores
     */
    public Page<Tutor> tutorFindAll(Integer page, Integer size, String sort, String direction) {
        return tutorService.findAll(page, size, sort, direction);
    }

    /**
     * Find tutor by id.
     *
     * @param id the id
     * @return the tutor
     */
    public Tutor tutorFindById(UUID id) {
        return tutorService.findById(id);
    }

    /**
     * Save tutor.
     *
     * @param tutor the tutor
     * @return the tutor
     */
    public Tutor tutorSave(Tutor tutor) {
        return tutorService.save(tutor);
    }

    /**
     * Search tutores.
     *
     * @param value     Nome, CPF ou RG
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of tutores.
     */
    public Page<Tutor> tutorSearch(String value, Integer page, Integer size, String sort, String direction) {
        return tutorService.search(value, page, size, sort, direction);
    }

    //////////////////////////////////////////////// USUÁRIO ////////////////////////////////////////////////

    /**
     * Find all usuarios.
     *
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of usuários
     */
    public Page<Usuario> usuarioFindAll(Integer page, Integer size, String sort, String direction) {
        return usuarioService.findAll(page, size, sort, direction);
    }

    /**
     * Find usuario by cpf.
     *
     * @param cpf the cpf
     * @return the usuario
     */
    public Usuario usuarioFindByCpf(String cpf) {
        return usuarioService.findByCpf(cpf);
    }

    /**
     * Find usuario by id.
     *
     * @param id the id
     * @return the usuario
     */
    public Usuario usuarioFindById(UUID id) {
        return usuarioService.findById(id);
    }

    /**
     * Save usuario.
     *
     * @param usuario the usuario
     * @return the usuario
     */
    public Usuario usuarioSave(Usuario usuario) {
        return usuarioService.save(usuario);
    }


    /**
     * Search usuários.
     *
     * @param value     Nome ou CPF
     * @param page      the page
     * @param size      the size
     * @param sort      the sort
     * @param direction the direction
     * @return the list of usuários
     */
    public Page<Usuario> usuarioSearch(String value, Integer page, Integer size, String sort, String direction) {
        return usuarioService.search(value, page, size, sort, direction);
    }
}

File: frontend/src/app/services/facade.service.ts
import { Injectable } from '@angular/core';

import { Animal } from '../entities/animal';
import { Atendimento } from '../entities/atendimento';
import { Exame } from '../entities/exame';
import { Tutor } from '../entities/tutor';
import { Usuario } from '../entities/usuario';
import { NotificationType } from '../enums/notification-type';
import { AgendamentoService } from './agendamento.service';
import { AnimalService } from './animal.service';
import { AtendimentoService } from './atendimento.service';
import { AuthService } from './auth.service';
import { ExameService } from './exame.service';
import { ImagemService } from './imagem.service';
import { NotificationService } from './notification.service';
import { TutorService } from './tutor.service';
import { UsuarioService } from './usuario.service';
import { Agendamento } from '../entities/agendamento';
import {Adocao} from "../entities/adocao";

@Injectable({
  providedIn: 'root'
})
export class FacadeService {

  constructor(
    private _agendamentoService: AgendamentoService,
    private _animalService: AnimalService,
    private _atendimentoService: AtendimentoService,
    private _authService: AuthService,
    private _exameService: ExameService,
    private _imagemService: ImagemService,
    private _notificationService: NotificationService,
    private _tutorService: TutorService,
    private _usuarioService: UsuarioService
  ) { }

  ////////////////////////////////////////////////// AGENDAMENTO //////////////////////////////////////////////////

  /**
   *
   * @param id
   * @returns
   */
  agendamentoDelete(id: string) {
    return this._agendamentoService.delete(id);
  }

  /**
   *
   * @returns
   */
  agendamentoFindAll(page: number, size: number, sort: string, direction: string) {
    return this._agendamentoService.findAll(page, size, sort, direction);
  }

  /**
   *
   * @param agendamento
   * @returns
   */
  agendamentoSave(agendamento: Agendamento) {
    return this._agendamentoService.save(agendamento);
  }

  /**
   *
   * @param value
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  agendamentoSearch(value: string, page: number, size: number, sort: string, direction: string) {
    return this._agendamentoService.search(value, page, size, sort, direction);
  }

  /**
   *
   * @param agendamento
   * @returns
   */
  agendamentoUpdate(agendamento: Agendamento) {
    return this._agendamentoService.update(agendamento);
  }

  ////////////////////////////////////////////////// ANIMAL //////////////////////////////////////////////////

  animalAdocaoDelete(id: string, idAdocao: string) {
    return this._animalService.adocaoDelete(id, idAdocao);
  }

  animalAdocaoFindAll(id: string, page: number, size: number, sort: string, direction: string) {
    return this._animalService.adocaoFindAll(id, page, size, sort, direction);
  }

  animalAdocaoSave(animal: Animal, adocao: Adocao) {
    return this._animalService.adocaoSave(animal, adocao);
  }

  animalAdocaoUpate(animal: Animal, adocao: Adocao) {
    return this._animalService.adocaoUpdate(animal, adocao);
  }

  /**
   *
   * @param id
   * @returns
   */
  animalDelete(id: string) {
    return this._animalService.delete(id);
  }

  /**
   *
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  animalFindAll(page: number, size: number, sort: string, direction: string) {
    return this._animalService.findAll(page, size, sort, direction);
  }

  /**
   *
   * @param id
   * @returns
   */
  animalFindById(id: string) {
    return this._animalService.findById(id);
  }

  /**
   *
   * @returns
   */
  animalGet() {
    return this._animalService.get();
  }

  /**
   *
   * @param animal
   * @param novaFoto
   * @returns
   */
  animalSave(animal: Animal, foto: any) {
    return this._animalService.save(animal, foto);
  }

  /**
   *
   * @param value
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  animalSearch(value: string, page: number, size: number, sort: string, direction: string) {
    return this._animalService.search(value, page, size, sort, direction);
  }

  /**
   *
   * @param animal
   */
  animalSet(animal: Animal) {
    this._animalService.set(animal);
  }

  /**
   *
   * @param animal
   * @param novaFoto
   * @param antigaFoto
   * @returns
   */
  animalUpdate(animal: Animal, novaFoto: any) {
    return this._animalService.update(animal, novaFoto);
  }

  ////////////////////////////////////////////////// ATENDIMENTO //////////////////////////////////////////////////

  /**
   *
   * @param id
   * @returns
   */
  atendimentoDelete(id: string) {
    return this._atendimentoService.delete(id);
  }

  /**
   *
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  atendimentoFindAll(page: number, size: number, sort: string, direction: string) {
    return this._atendimentoService.findAll(page, size, sort, direction);
  }

  /**
   *
   * @param atendimento
   * @param documentosToSave
   * @returns
   */
  atendimentoSave(atendimento: Atendimento, documentosToSave: Array<File> | null) {
    return this._atendimentoService.save(atendimento, documentosToSave);
  }

  /**
   *
   * @param value
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  atendimentoSearch(value: string, page: number, size: number, sort: string, direction: string) {
    return this._atendimentoService.search(value, page, size, sort, direction);
  }

  /**
   *
   * @param atendimento
   * @param documentosToSave
   * @param documentosToDelete
   * @returns
   */
  atendimentoUpdate(atendimento: Atendimento, documentosToSave: Array<File> | null, documentosToDelete: Array<string> | null) {
    return this._atendimentoService.update(atendimento, documentosToSave, documentosToDelete);
  }

  ////////////////////////////////////////////////// AUTHENTICATION //////////////////////////////////////////////////

  /**
    *
    * @returns
    */
  authIsAuthenticated() {
    return this._authService.isAuthenticated();
  }

  /**
   *
   * @param user
   * @returns
   */
  authLogin(user: any) {
    return this._authService.login(user);
  }

  /**
   *
   * @returns
   */
  authLogout() {
    return this._authService.logout();
  }

  /**
   *
   * @returns
   */
  authGetCurrentUser() {
    return this._authService.getCurrentUser();
  }

  /**
   *
   * @param currentUser
   * @returns
   */
  authSetCurrentUser(currentUser: any) {
    return this._authService.setCurrentUser(currentUser);
  }

  ////////////////////////////////////////////////// EXAME //////////////////////////////////////////////////

  /**
   *
   * @param id
   * @returns
   */
  exameDelete(id: string) {
    return this._exameService.delete(id);
  }


  /**
   *
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  exameFindAll(page: number, size: number, sort: string, direction: string) {
    return this._exameService.findAll(page, size, sort, direction);
  }

  /**
   *
   * @param exame
   * @returns
   */
  exameSave(exame: Exame) {
    return this._exameService.save(exame);
  }

  /**
   *
   * @param value
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  exameSearch(value: string, page: number, size: number, sort: string, direction: string) {
    return this._exameService.search(value, page, size, sort, direction);
  }

  /**
   *
   * @param exame
   * @returns
   */
  exameUpdate(exame: Exame) {
    return this._exameService.update(exame);
  }

  ////////////////////////////////////////////////// IMAGEM //////////////////////////////////////////////////

  /**
   *
   * @param imagem
   * @returns
   */
  imagemToBase64(imagem: any) {
    return this._imagemService.toBase64(imagem);
  }

  ////////////////////////////////////////////////// NOTIFICATION //////////////////////////////////////////////////

  /**
   *
   * @param message
   * @param type
   */
  notificationShowNotification(message: string, type: NotificationType) {
    this._notificationService.show(message, type);
  }

  ////////////////////////////////////////////////// TUTOR //////////////////////////////////////////////////

  /**
   *
   * @param id
   * @returns
   */
  tutorDelete(id: string) {
    return this._tutorService.delete(id);
  }

  /**
   *
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  tutorFindAll(page: number, size: number, sort: string, direction: string) {
    return this._tutorService.findAll(page, size, sort, direction);
  }

  /**
   *
   * @param id
   * @returns
   */
  tutorFindById(id: string) {
    return this._tutorService.findById(id);
  }

  /**
   *
   * @returns
   */
  tutorGet() {
    return this._tutorService.get();
  }

  /**
   *
   * @param tutor
   * @param novaFoto
   * @returns
   */
  tutorSave(tutor: Tutor, novaFoto: any) {
    return this._tutorService.save(tutor, novaFoto);
  }

  /**
   *
   * @param value
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  tutorSearch(value: string, page: number, size: number, sort: string, direction: string) {
    return this._tutorService.search(value, page, size, sort, direction);
  }

  /**
   *
   * @param tutor
   */
  tutorSet(tutor: Tutor) {
    this._tutorService.set(tutor);
  }

  /**
   *
   * @param tutor
   * @param novaFoto
   * @param antigaFoto
   * @returns
   */
  tutorUpdate(tutor: Tutor, novaFoto: any, antigaFoto: any) {
    return this._tutorService.update(tutor, novaFoto, antigaFoto);
  }

  ////////////////////////////////////////////////// USUÁRIO //////////////////////////////////////////////////

  /**
   *
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  usuarioFindAll(page: number, size: number, sort: string, direction: string) {
    return this._usuarioService.findAll(page, size, sort, direction);
  }

  /**
   *
   * @param id
   * @returns
   */
  usuarioFindById(id: string) {
    return this._usuarioService.findById(id);
  }

  /**
   *
   * @returns
   */
  usuarioGet() {
    return this._usuarioService.get();
  }

  /**
   *
   * @param usuario
   * @param novaFoto
   * @param antigaFoto
   * @returns
   */
  usuarioSave(usuario: Usuario, novaFoto: any) {
    return this._usuarioService.save(usuario, novaFoto);
  }

  /**
   *
   * @param value
   * @param page
   * @param size
   * @param sort
   * @param direction
   * @returns
   */
  usuarioSearch(value: string, page: number, size: number, sort: string, direction: string) {
    return this._usuarioService.search(value, page, size, sort, direction);
  }

  /**
   *
   * @param usuario
   */
  usuarioSet(usuario: Usuario) {
    this._usuarioService.set(usuario);
  }

  /**
   *
   * @param usuario
   * @param novaFoto
   * @param antigaFoto
   * @returns
   */
  usuarioUpdate(usuario: Usuario, novaFoto: any, antigaFoto: any) {
    return this._usuarioService.update(usuario, novaFoto, antigaFoto);
  }
}
